chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -113,74 +107,204 @@ pub fn validate_prepared_transactions(
+            // proofs.
+            1 => blocks_after_last_last_chunk.push(position.prev_block.clone()),
+            // If we have seen the 2nd chunk, we are done.
+            2 => break,
+            _ => unreachable!("chunks_seen should never exceed 2"),
         }
-        if prev_chunks_seen == 2 || is_genesis {
+
+        if position.prev_block.header().is_genesis() {
             break;
         }
 
-        block_hash = prev_hash;
-        (current_shard_id, current_shard_index) =
-            epoch_manager.get_prev_shard_id(&prev_hash, current_shard_id)?;
+        let prev_prev_block = store.get_block(&prev_prev_hash)?;
+        position = TraversalPosition {
+            shard_id: prev_shard_id,
+            prev_hash: prev_prev_hash,
+            prev_block: prev_prev_block,
+            chunks_seen: chunks_seen_update,
+            // If we have not seen chunks, switch to previous shard id, but
+            // once we just saw the first chunk, start keeping its shard id.
+            last_chunk_shard_id: if position.chunks_seen == 0 {
+                prev_shard_id
+            } else {
+                position.last_chunk_shard_id
+            },
+            last_chunk_shard_index: if position.chunks_seen == 0 {
+                prev_shard_index
+            } else {
+                position.last_chunk_shard_index
+            },
+        };
     }
 
+    implicit_transition_params.reverse();
     Ok(StateWitnessBlockRange {
-        blocks_after_last_chunk,
+        implicit_transition_params,
         blocks_after_last_last_chunk,
-        last_chunk_shard_id,
-        last_chunk_shard_index,
+        last_chunk_shard_id: position.last_chunk_shard_id,
+        last_chunk_shard_index: position.last_chunk_shard_index,
     })
 }
 
+/// Checks if chunk validation requires a transition to new shard layout in the
+/// block with `prev_hash`, with a split resulting in the `shard_uid`, and if
+/// so, returns the corresponding resharding transition parameters.
+fn check_shard_layout_transition(
+    epoch_manager: &dyn EpochManagerAdapter,
+    prev_hash: &CryptoHash,
+    shard_layout: &ShardLayout,
+    shard_uid: ShardUId,
+    chunks_seen: u32,
+) -> Result<Option<ImplicitTransitionParams>, Error> {
+    // If we have already seen a new chunk, we don't need to validate
+    // resharding transition.
+    if chunks_seen > 0 {
+        return Ok(None);
+    }
+
+    let prev_epoch_id = epoch_manager.get_prev_epoch_id_from_prev_block(prev_hash)?;
+    let prev_shard_layout = epoch_manager.get_shard_layout(&prev_epoch_id)?;
+    let block_has_new_shard_layout =
+        epoch_manager.is_next_block_epoch_start(prev_hash)? && shard_layout != &prev_shard_layout;
+
+    if !block_has_new_shard_layout {
+        return Ok(None);
+    }
+
+    let Some(ReshardingEventType::SplitShard(params)) =
+        ReshardingEventType::from_shard_layout(shard_layout, *prev_hash)?
+    else {
+        return Ok(None);
+    };
```

Comment:
mini nit: I think a match statement on the ReshardingEventType will be more future proof so that the compiler warns that a new event is not handled here. let else return is fine for handling Option though.